db-canister/db.mo

```diff
@@ -0,0 +1,136 @@
+    func castData(data : BTree.Data<PK1, Nat>) : BTree.Data<PK2, Nat> = {
+      var count = data.count;
+      kvs = Array.tabulateVar<?(PK2, Nat)>(
+        data.kvs.size(),
+        func(n) = switch (data.kvs[n]) { case (null) { null }; case (?k) { ?(castPk(k.0), k.1) } },
+      );
+    };
+    func castIndexNodeRecv(node : BTree.Node<PK1, Nat>) : BTree.Node<PK2, Nat> = switch (node) {
+      case (#leaf x) #leaf({ data = castData(x.data) });
+      case (#internal x) #internal({
+        children = Array.tabulateVar<?BTree.Node<PK2, Nat>>(
+          x.children.size(),
+          func(n) = switch (x.children[n]) {
+            case (null) { null };
+            case (?k) { ?castIndexNodeRecv(k) };
+          },
+        );
+        data = castData(x.data);
+      });
+    };
+    {
+      db = {
+        var reuse_queue = src.db.reuse_queue;
+        vec = Vector.map<?T1, ?T2>(
+          src.db.vec,
+          func(x) = switch (x) {
+            case (?item) ?cast(item);
+            case (null) null;
+          },
+        );
+      };
+      pk = {
+        order = src.pk.order;
+        var size = src.pk.size;
+        var root = castIndexNodeRecv(src.pk.root);
+      };
+      updatedAt = src.updatedAt;
+    };
+  };
+
+  public type DbInit<T, PK> = {
+    db : RXMDB.RXMDB<T>;
+    pk : PK.Init<PK>;
+    updatedAt : IDX.Init<Nat64>;
+  };
+
+  public type DbUse<T, PK> = {
+    db : RXMDB.Use<T>;
+    pk : PK.Use<PK, T>;
+    updatedAt : IDX.Use<Nat64, T>;
+  };
+
+  public func empty<T, PK>() : DbInit<T, PK> = {
+    db = RXMDB.init<T>();
+    pk = PK.init<PK>(?32);
+    updatedAt = IDX.init<Nat64>(?32);
+  };
+
+  public func use<T, PK>(
+    init : DbInit<T, PK>,
+    pkGet : (item : T) -> PK,
+    pkCompare : (pk1 : PK, pk2 : PK) -> { #less; #equal; #greater },
+    updatedAtGet : (item : T) -> Nat32,
+  ) : DbUse<T, PK> {
+    let obs = RXMDB.init_obs<T>();
+    // PK
+    let pk_config : PK.Config<PK, T> = {
+      db = init.db;
+      obs;
+      store = init.pk;
+      compare = pkCompare;
+      key = pkGet;
+      regenerate = #no;
+    };
+    PK.Subscribe<PK, T>(pk_config);
+
+    let updatedAt_config : IDX.Config<Nat64, T> = {
+      db = init.db;
+      obs;
+      store = init.updatedAt;
+      compare = Nat64.compare;
+      key = func(idx : Nat, d : T) = ?((Nat64.fromNat(Nat32.toNat(updatedAtGet(d))) << 32) | Nat64.fromNat(idx));
+      regenerate = #no;
+      keep = #all;
+    };
+    IDX.Subscribe(updatedAt_config);
+
+    return {
+      db = RXMDB.Use<T>(init.db, obs);
+      pk = PK.Use(pk_config);
+      updatedAt = IDX.Use(updatedAt_config);
+    };
+  };
+
+  public func pushUpdates<T, PK>(use : DbUse<T, PK>, docs : [T]) : [T] {
+    // var conflicts : Vector.Vector<TodoListItemDoc> = Vector.new();
+    for (doc in docs.vals()) {
+      use.db.insert(doc);
+      // let ?rec = callerDb.pk.get(doc.id) else Debug.trap("Not found");
```

Comment:
I feel that we might need to ask for @timohanke guidance here, as I lack the knowledge to understand this code.
I'd copy it over form https://github.com/research-ag/wallet/pull/3

Reply:
The commented variable `conflicts` is defined as `Vector<TodoListItemDoc>`. The type `TodoListItemDoc` seems to come from a previous RXDB example that was used as an example in early development. So all this may just be old code. 

Do we have a mechanism for conflict resolution elsewhere? Or do we even need conflict resolution?

Reply:
I honestly don't know. What exactly does it need to be validated here?
What's the meaning of a "conflict resolution" in this case?

Reply:
currently is not needed a conflict resolution. on this case. i think is possible to delete the commented code. in case of needed is on the PR or original branch. so no need for it here!